river, rivertile: implement attach-mode aside.

0xfea5 opened the pull request:
See: https://github.com/riverwm/river/issues/904

MaxVerevkin commented:
Maybe instead of introducing the concept of "main count" in the river itself, the attach mode can be called something like "nth" and be parametrized by an integer? Then you or your layout generator may choose to set the attach mode every time the "main count" changes.

```sh
riverctl attach-mode nth $N
```

Leon-Plickat commented on river/Config.zig:
```diff
@@ -34,6 +34,7 @@ const View = @import("View.zig");
 pub const AttachMode = enum {
     top,
     bottom,
+    aside,
```
There should be a single name for this. I like either 'aside' or 'n-th'.

Leon-Plickat commented on river/View.zig:
```diff
@@ -36,6 +36,8 @@ const XwaylandView = @import("XwaylandView.zig");
 
 const log = std.log.scoped(.view);
 
+pub var nMainCount: u64 = 1;
```
This is the wrong place to store this, IMO.

It'd be better to convert `AttachMode` from `enum` to `union(enum)`, which can then hold the index directly.

```zig
pub const AttachMode = union(enum) {
    top: void,
    bottom: void,
    aside: usize,
};
```

Also the attach mode would then need to be a per-output configuration, which means instead of a single `attach-mode` command there would need to be two commands, `output-attach-mode` and `default-attach-mode`, similar to how layout namespaces work.

Leon-Plickat commented on river/command/layout.zig:
```diff
@@ -83,3 +84,18 @@ pub fn sendLayoutCmd(
     if (layout == output.layout) server.root.applyPending();
 }
+
+/// Called by the layout manager to update the number of Views in the main area.
```
This command may also be called by the user directly.
The concept of main count should not exist in river, only the layout generator.
This should be folded into the `attach-mode` command:
```sh
riverctl attach-mode aside 2
```

Leon-Plickat commented on rivertile/main.zig:
```diff
@@ -179,6 +180,37 @@ const Output = struct {
                                 if (arg >= 1) output.main_count = @intCast(arg);
                             },
                         }
+
+                        var mainCount: [32:0]u8 = undefined;
+                        _ = std.fmt.bufPrintZ(&mainCount, "{d}", .{output.main_count}) catch {
+                            std.log.err("could not write {d} to buf", .{output.main_count});
+                            return;
+                        };
+
+                        const args = &[_][*:0]const u8{
+                            "set-main-count",
+                            &mainCount,
+                        };
+
+                        riverctl.run(args) catch |err| {
+                            switch (err) {
+                                error.ConnectFailed => {
+                                    std.log.err("failed to connect to wayland compositor", .{});
+                                },
+                                error.RiverControlNotAdvertised => {
+                                    std.log.err("river control not advertised", .{});
+                                },
+                                error.SeatNotAdverstised => {
+                                    std.log.err("seat not advertised", .{});
+                                },
+                                error.CannotGetRegistry => {
+                                    std.log.err("failed to get registry", .{});
+                                },
+                                error.RunCommandFailed => {
+                                    std.log.err("failed to run command", .{});
+                                },
+                            }
+                        };
```
I am pretty sure this is out of scope for rivertile and will not be merged.

Rivertile is supposed to be a reference implementation with a very restricted feature set and has no ambitions to serve the use cases of a wide range of users. Users running (and creating) other layout generators is highly encouraged.

Leon-Plickat commented on rivertile/riverctl.zig:
```diff
@@ -0,0 +1,81 @@
+    callback.setListener(?*anyopaque, callbackListener, null);
+
+    // Loop until our callback is called and we exit.
+    while (display.dispatch() != .SUCCESS) {
+        std.log.err("failed to dispatch wayland events", .{});
+    }
+}
+
+fn registryListener(registry: *wl.Registry, event: wl.Registry.Event, ctx: *Context) void {
+    switch (event) {
+        .global => |global| {
+            if (mem.orderZ(u8, global.interface, wl.Seat.getInterface().name) == .eq) {
+                assert(ctx.seat == null); // TODO: support multiple seats
+                ctx.seat = registry.bind(global.name, wl.Seat, 1) catch @panic("out of memory");
+            } else if (mem.orderZ(u8, global.interface, zriver.ControlV1.getInterface().name) == .eq) {
+                ctx.control = registry.bind(global.name, zriver.ControlV1, 1) catch @panic("out of memory");
+            }
+        },
+        .global_remove => {},
+    }
+}
+
+fn callbackListener(_: *zriver.CommandCallbackV1, event: zriver.CommandCallbackV1.Event, _: ?*anyopaque) void {
+    switch (event) {
+        .success => |success| {
+            if (mem.len(success.output) > 0) {
+                const stdout = io.getStdOut().writer();
+                stdout.print("{s}\n", .{success.output}) catch @panic("failed to write to stdout");
+            }
+        },
+        .failure => |failure| {
+            // A small hack to provide usage text when river reports an unknown command.
+            if (mem.orderZ(u8, failure.failure_message, "unknown command") == .eq) {
+                std.log.err("unknown command", .{});
+                return;
+            }
+            std.log.err("{s}", .{failure.failure_message});
+        },
+    }
+}
```
In the unlikely event that Isaac wants this feature in rivertile, this would be the wrong way to implement it.

Why create a new wayland connection when we can just use the one we already have for the layout protocol?

Leon-Plickat commented on river/command.zig:
```diff
@@ -69,6 +69,9 @@ const command_impls = std.ComptimeStringMap(
         .{ "map",                       @import("command/map.zig").map },
         .{ "map-pointer",               @import("command/map.zig").mapPointer },
         .{ "map-switch",                @import("command/map.zig").mapSwitch },
+        // NOTE: `set-main-count` should only be called by the layout manager and not directly.
+        // Advertises the number of views in the main area to river.
```
```suggestion
```

There is absolutely no reason users should not use this command directly. In fact I suspect layout generators calling it will be the minority.

Leon-Plickat commented on river/View.zig:
```diff
@@ -421,6 +423,14 @@ pub fn setPendingOutput(view: *Self, output: *Output) void {
         .top => output.pending.wm_stack.prepend(view),
         .bottom => output.pending.wm_stack.append(view),
+        .aside => {
+            attach_aside(
+                &output.pending.wm_stack,
+                output.pending.tags,
+                &view.pending_wm_stack_link,
+                .pending,
+            );
+        },
```
If you make the function take `self: *View` and `output: *Output` as arguments, it would be a bit cleaner IMO.

0xfea5 replied:
Your suggestion was what I initially did, however there is another place (View.zig:575) where `attach_aside()` is called without an `output` to be passed as a parameter. I don't like the way it looks either so If you have any better ideas I'd be glad to see.

0xfea5 commented:
Thank you both for the comments, I will take some time to revision the changes and make another PR.

ifreund reviewed (changes requested):
I have not looked at any of the code here in detail. There are some high level changes that need to be made though before I do a thorough review

1. Make this not a breaking change.
2. Drop all the rivertile changes, rivertile is feature complete.
3. Document the new feature in the riverctl man page.

0xfea5 commented:
@ifreund @Leon-Plickat

I have pushed my most recent commits. I kept the changes simple and I believe the patch is in a better state now. The only part it can may be improved is for `View:attach_aside()` to have a cleaner signature, as Leon initially suggested.

ifreund commented on doc/riverctl.1.scd:
```diff
@@ -347,8 +347,17 @@ matches everything while _\*\*_ and the empty string are invalid.
 
-*attach-mode* *top*|*bottom*
-	Configure where new views should attach to the view stack.
+*default-attach-mode* *top*|*bottom*|*aside <N>*
```
This rename is still a breaking change and needs to be reverted

0xfea5 replied:
I am thinking about adding `attach-mode` as an alias to `default-attach-mode` to keep both to keep both consistency (with options like default-layout/output-layout) as well as compatibility with current configurations. What do you think?

ifreund replied:
Sure, an alias would be fine.

0xfea5 replied:
FYI, the change is up. You may review everything if you haven't already.

ifreund commented on river/View.zig:
```diff
@@ -476,6 +485,30 @@ pub fn applyConstraints(self: *Self, box: *wlr.Box) void {
 }
 
+/// Attach current view after n Views on the stack
+pub fn attach_aside(
+    view: *Self,
+    comptime state_field: @Type(.EnumLiteral),
+    comptime state_link_field: @Type(.EnumLiteral),
+    wm_stack: anytype,
+    tags: usize,
+    n: usize,
+) void {
```
This level of metaprogramming should not be necessary. I just pushed a small refactor to master that should allow this function to only deal with the pending wm_stack.

0xfea5 replied:
That factor does indeed help. Perhaps we can move the `tags` inside `pending` to keep similar structure as `Output.pending`?

ifreund replied:
I'm not sure what `tags` you are referring to here, could you be more precise?

0xfea5 replied:
I am referring to the two structs `Root.fallback.pending` and `Output.pending`. Seeing you have refactored `Root.fallback.inflight` away we can move `Root.fallback.tags` in `Root.fallback.pending` like this

**Root.zig:**
```zig
fallback: struct {
...
  pending: struct {
     tags: u32 = 1 << 0,
     focus_stack: wl.list.Head(View, .pending_focus_stack_link),
     wm_stack: wl.list.Head(View, .pending_wm_stack_link),
  },
...
},
```

That has the benefit of keeping the same struct as

**Output.zig:**
```zig
pending: struct {
    tags: u32 = 1 << 0,
    focus_stack: wl.list.Head(View, .pending_focus_stack_link),
    wm_stack: wl.list.Head(View, .pending_wm_stack_link),
},
```

while also simplifying the call to `attach_aside()` since we will only have to pass `Output.pending`/`Root.fallback.pending` depending on the situation.

ifreund replied:
Ah, yeah making those the same type is a good idea. Just pushed 6e9bd83e1d0e2132d2def338b51f66867cd88907

ifreund commented on doc/riverctl.1.scd:
```diff
@@ -347,8 +347,20 @@ matches everything while _\*\*_ and the empty string are invalid.
 
-*attach-mode* *top*|*bottom*
-	Configure where new views should attach to the view stack.
+*attach-mode* *top*|*bottom*|*aside <N>*
+	Alias to default-attach-mode.
```
I'd prefer a mention in the description of `default-attach-mode` such as "attach-mode may be used as an alias for this command for backwards compatibility"

ifreund commented:
Also, could we rename this to something other than "aside"? The word doesn't really make sense in the context of inserting to a list at an arbitrary index. I think @MaxVerevkin's suggestion of "attach-mode nth 5" is more clear, as is "attach-mode after 5" or "attach-mode index 5"

0xfea5 commented:
I've rebased and pushed my latest changes. I renamed `attach_aside` to `attach_after` for better clarity as well as refactor the parts of the codes that @ifreund suggested. Since there is no inflight state at the moment, `attach_after` is hard-coded to work with the pending state.

It seems that all the changes from `master` are included in my latest commit. I am not sure if that's the default behavior of rebase or if it's a mistake on my part. Hopefully, it won't cause much trouble when merging.

ifreund reviewed (approved):
LGTM, thanks for the patch!

> It seems that all the changes from master are included in my latest commit. I am not sure if that's the default behavior of rebase or if it's a mistake on my part. Hopefully, it won't cause much trouble when merging.

Everything looks fine to me, I just added a commit with some final style tweaks of my own and squashed everything up :)

0xfea5 commented:
Good to know everything worked well at the end. Thank you to everyone who made suggestions and/or comments, it helped me a lot.
